### PR TITLE
test: Replica fork test

### PR DESCRIPTION
**Pull request by anna-carroll**
## PR Checklist

- [ ] Added Tests
- [ ] Updated Documentation
- [ ] Updated CHANGELOG.md for the appropriate package

### CODE REVIEW

**Comment by anna-carroll** [comment]
Note: currently not working because Harness functions don't work in the fork environment (`setCommittedRoot` and `setMessageStatus`) - need to use vm.etch to replace Replica code with ReplicaHarness code, OR, use cheatcodes to implement an equivalent of `setCommittedRoot` and `setMessageStatus`

**Comment by odyslam** [comment]
```
  // At commit: 4679f48f0f7392849e75a17487c4bfd6b9d08f33, this is the storage layout of Replica:
  // FOUNDRY_PROFILE=core forge inspect Replica --pretty storage-layout
  // +-------------------+-----------------------------+------+--------+-------+
  // | Name              | Type                        | Slot | Offset | Bytes |
  // +=========================================================================+
  // | _initialized      | bool                        | 0    | 0      | 1     |
  // |-------------------+-----------------------------+------+--------+-------|
  // | _initializing     | bool                        | 0    | 1      | 1     |
  // |-------------------+-----------------------------+------+--------+-------|
  // | __gap             | uint256[50]                 | 1    | 0      | 1600  |
  // |-------------------+-----------------------------+------+--------+-------|
  // | _owner            | address                     | 51   | 0      | 20    |
  // |-------------------+-----------------------------+------+--------+-------|
  // | __gap             | uint256[49]                 | 52   | 0      | 1568  |
  // |-------------------+-----------------------------+------+--------+-------|
  // | updater           | address                     | 101  | 0      | 20    |
  // |-------------------+-----------------------------+------+--------+-------|
  // | state             | enum NomadBase.States       | 101  | 20     | 1     |
  // |-------------------+-----------------------------+------+--------+-------|
  // | committedRoot     | bytes32                     | 102  | 0      | 32    |
  // |-------------------+-----------------------------+------+--------+-------|
  // | __GAP             | uint256[47]                 | 103  | 0      | 1504  |
  // |-------------------+-----------------------------+------+--------+-------|
  // | remoteDomain      | uint32                      | 150  | 0      | 4     |
  // |-------------------+-----------------------------+------+--------+-------|
  // | optimisticSeconds | uint256                     | 151  | 0      | 32    |
  // |-------------------+-----------------------------+------+--------+-------|
  // | entered           | uint8                       | 152  | 0      | 1     |
  // |-------------------+-----------------------------+------+--------+-------|
  // | confirmAt         | mapping(bytes32 => uint256) | 153  | 0      | 32    |
  // |-------------------+-----------------------------+------+--------+-------|
  // | messages          | mapping(bytes32 => bytes32) | 154  | 0      | 32    |
  // |-------------------+-----------------------------+------+--------+-------|
  // | __GAP             | uint256[45]                 | 155  | 0      | 1440  |
  // +-------------------+-----------------------------+------+--------+-------+

  /// @notice Set the commited Root to a deployed Replica, by manipulating storage
  function setReplicaCommittedRoot(bytes32 root) public {
    bytes32 committedRootSlot = bytes32(uint256(102));
    bytes32 confirmAtStartSlot = keccak256(
      abi.encodePacked(root, bytes32(uint256(153)))
    );
    vm.store(address(replicaProxy), committedRootSlot, root);
    // How mappings are placed in storage
    // https://docs.soliditylang.org/en/v0.8.13/internals/layout_in_storage.html
    vm.store(address(replicaProxy), confirmAtStartSlot, bytes32(uint256(1)));
    assertEq(replicaProxy.committedRoot(), root);
    assertEq(replicaProxy.confirmAt(root), 1);
  }
}
```

**Comment by odyslam** [comment]
We shouldn't use etch, as we want to test the actual on-chain code imho.

File: packages/contracts-core/contracts/test/Replica.t.sol [hunk 1]
```diff
@@ -453,9 +453,9 @@ contract ReplicaTest is ReplicaHandlers {
 
     function test_setOptimisticTimeoutOnlyOwner() public {
```
Comment (odyslam) [inline packages/contracts-core/contracts/test/Replica.t.sol:454]: Ideally, we should revisit this and fuzz it, but good choice for something more realistic.

File: packages/contracts-ops/contracts/scripts/InitializeImplementations.s.sol [hunk 2]
```diff
@@ -56,26 +51,20 @@ contract InitializeImplementationsLogic is Script, Config {
             payable(bridgeRouterUpgrade(_domain).implementation)
         );
         if (address(bridgeRouter.tokenRegistry()) == address(0)) {
-            console2.log("   initialize bridgeRouter ", _domain);
```
Comment (odyslam) [inline packages/contracts-ops/contracts/scripts/InitializeImplementations.s.sol:59]: Why did we remove the logs again?

File: packages/contracts-ops/contracts/scripts/UpgradeCallBatches.s.sol [hunk 3]
```diff
@@ -21,32 +21,23 @@ contract UpgradeCallBatchLogic is Script, Config, CallBatch {
...
         );
     }
 
-    function pushSingleUpgrade(
-        Upgrade memory _upgrade,
-        string memory contractName
-    ) private {
+    function pushSingleUpgrade(Upgrade memory _upgrade) internal {
         // check if upgrade is unnecessary
         (, bytes memory result) = address(_upgrade.beacon).call("");
```
Comment (odyslam) [inline packages/contracts-ops/contracts/scripts/UpgradeCallBatches.s.sol:36]: Should this be a call (wrt broadcast)? Perhaps easier to do the vm.load() I mentioned elsewhere

**Review by odyslam — APPROVED** [review]

File: packages/contracts-ops/contracts/test/Reboot.t.sol [hunk 4]
```diff
@@ -4,21 +4,71 @@ pragma abicoder v2;
...
+        // setup replica to actual value
+        replica = ReplicaHarness(
+            address(getReplicaOf(localDomainName, remote))
+        );
+        // set committed root and optimistic timeout
+        committedRoot = replica.committedRoot();
+        optimisticTimeout = replica.optimisticSeconds();
+        // setup fake app handlers
+        setUpBadHandlers();
+        // Upgrade Replica to Replica Harness
```
Comment (prestwich) [inline packages/contracts-ops/contracts/test/Reboot.t.sol:49]: 👍  on approach

File: packages/contracts-ops/contracts/test/Reboot.t.sol [hunk 5]
```diff
@@ -4,21 +4,71 @@ pragma abicoder v2;
...
             address(getGovernanceRouter(_domain)),
             getDomainNumber(_domain)
         );
+        // set remote domain to actual values
+        remote = getConnections(_domain)[0];
+        remoteDomain = getDomainNumber(remote);
+        // set home domain to actual values
+        homeDomain = getDomainNumber(_domain);
+        // setup replica to actual value
+        replica = ReplicaHarness(
```
Comment (odyslam) [inline packages/contracts-ops/contracts/test/Reboot.t.sol:41]: This doesn't seem right. 

The address is of a replica contract, so the harness functions shouldn't exist.

**Comment by prestwich** [comment]
@odyslam the reason that i'd prefer the harness is that it doesn't require a reference document to know what it's doing. It doesn't change the bytecode that we're testing, just extends it

File: packages/contracts-ops/contracts/test/Reboot.t.sol [hunk 6]
```diff
@@ -4,21 +4,71 @@ pragma abicoder v2;
...
+            address(getReplicaOf(localDomainName, remote))
+        );
+        // set committed root and optimistic timeout
+        committedRoot = replica.committedRoot();
+        optimisticTimeout = replica.optimisticSeconds();
+        // setup fake app handlers
+        setUpBadHandlers();
+        // Upgrade Replica to Replica Harness
+        // deploy ReplicaHarness implementation
+        ReplicaHarness harnessImpl = new ReplicaHarness(
```
Comment (odyslam) [inline packages/contracts-ops/contracts/test/Reboot.t.sol:51]: Okay, now I see. 

I really think that fork tests should be **identical** to the on-chain reality, with minimum changes. I think it's much better to do vm.load() and change a single storage slot than upgrading Replica to a harness.

Comment (odyslam) [reply packages/contracts-ops/contracts/test/Reboot.t.sol:51]: I can do that in a later PR.

**Review by odyslam — APPROVED** [review]

**Pull request merged**